nomad/node_endpoint.go

```diff
@@ -486,8 +488,7 @@ func (n *Node) UpdateStatus(args *structs.NodeUpdateStatusRequest, reply *struct
 	}
 
 	// Check if we should trigger evaluations
-	transitionToReady := transitionedToReady(args.Status, node.Status)
-	if structs.ShouldDrainNode(args.Status) || transitionToReady || args.Status == structs.NodeStatusDisconnected {
+	if shouldCreateNodeEval(originalNode, node) {
```

Comment:
I think we took this refactoring a little too far.

`Register` and `UpdateStatus` are distinct because register includes the full fingerprint. Folding the check for status updates together with the check for the register code path has several disadvantages:
* We're doing a bunch of resources checks that are meaningless because they'll never have changed as part of an `UpdateStatus` call
* We have to copy the `Node`, which is extra allocations. (And I'm not even sure why a check for changes mutates the object?)
* We've introducing implicit coupling between fingerprinting and `UpdateStatus` that didn't exist before and can introduce future bugs if we want to change the fingerprinting conditionals.

Reply:
Gotcha. Exactly the feedback I was looking for. I'll back this up a level.